File: backend/tax_center.py
"""Indonesian Tax Center — deterministic estimates. AI only explains.
Regimes: PPh Final UMKM 0,5% (PP 55/2022) vs Tarif Umum PPh Badan 22% (fasilitas Pasal 31E).
PPN opsional (PKP 12% / non-PKP). PPh 21 (gaji) & PPh 23 (jasa/sewa)."""
import financial_engine as fe

PTKP_TK0 = 54_000_000            # PTKP Tidak Kawin / 0 tanggungan
FACILITY_LIMIT = 4_800_000_000   # batas peredaran bruto fasilitas Pasal 31E
PPH_FINAL_RATE = 0.005
PPN_RATE = 0.12
PPH21_RATE = 0.05
PPH23_RATE = 0.02
BADAN_RATE = 0.22


def compute(lines, pkp: bool = False):
    pnl_all = fe.profit_and_loss(lines)
    pnl = pnl_all["current"]
    bal = fe._acct_balances(lines)
    revenue = pnl["revenue"]
    net = pnl["net_profit"]

    # --- PPh Final UMKM 0,5% ---
    pph_final_annual = round(revenue * PPH_FINAL_RATE, 2)
    monthly_final = [{"month": s["month"], "revenue": s["revenue"],
                      "amount": round(s["revenue"] * PPH_FINAL_RATE, 2)} for s in pnl_all["series"]]

    # --- PPh Badan Tarif Umum (Pasal 31E) ---
    facility = revenue <= FACILITY_LIMIT
    if facility:
        badan = round(net * BADAN_RATE * 0.5, 2)
        eff = round(BADAN_RATE * 0.5 * 100, 2)
    else:
        pf = net * (FACILITY_LIMIT / revenue)
        badan = round(pf * (BADAN_RATE * 0.5) + (net - pf) * BADAN_RATE, 2)
        eff = round(badan / net * 100, 2) if net else 0
    badan_full = round(net * BADAN_RATE, 2)

    cheaper = "final" if pph_final_annual <= badan else "badan"
    saving = round(abs(pph_final_annual - badan), 2)

    # --- PPN ---
    material_base = round(bal.get("5000", 0), 2)  # pembelian material sbg dasar PPN masukan
    if pkp:
        ppn_out = round(revenue * PPN_RATE, 2)
        ppn_in = round(material_base * PPN_RATE, 2)
        ppn_payable = round(ppn_out - ppn_in, 2)
    else:
        ppn_out = ppn_in = ppn_payable = 0.0

    # --- PPh 21 (gaji) ---
    payroll = round(bal.get("6000", 0), 2)
    taxable21 = max(0.0, payroll - PTKP_TK0)
    pph21 = round(taxable21 * PPH21_RATE, 2)

    # --- PPh 23 (sewa/jasa) ---
    rent = round(bal.get("6200", 0), 2)
    pph23_withhold = round(rent * PPH23_RATE, 2)            # yang wajib dipotong atas sewa
    pph23_credit = round(revenue * PPH23_RATE, 2)           # estimasi dipotong pelanggan (kredit pajak)

    chosen_annual = pph_final_annual if cheaper == "final" else badan
    obligations = [
        {"tax": "PPh Final UMKM 0,5%", "amount": round(pph_final_annual / 12, 2),
         "freq": "Bulanan", "due": "Tgl 15 bulan berikutnya", "type": "final"},
    ]
    if pkp:
        obligations.append({"tax": "PPN (SPT Masa PPN)", "amount": round(ppn_payable / 12, 2) if ppn_payable else 0,
                            "freq": "Bulanan", "due": "Akhir bulan berikutnya", "type": "ppn"})
    if pph21 > 0:
        obligations.append({"tax": "PPh 21 Karyawan", "amount": round(pph21 / 12, 2),
                            "freq": "Bulanan", "due": "Tgl 10 bulan berikutnya", "type": "pph21"})
    obligations.append({"tax": "SPT Tahunan Badan", "amount": chosen_annual,
                        "freq": "Tahunan", "due": "4 bulan setelah tutup buku", "type": "annual"})

    return {
        "revenue": revenue, "net_profit_before_tax": net,
        "pph_final": {"rate_pct": 0.5, "base": revenue, "annual": pph_final_annual,
                      "monthly_avg": round(pph_final_annual / 12, 2), "monthly": monthly_final},
        "pph_badan": {"rate_pct": 22, "facility_31e": facility, "effective_rate_pct": eff,
                      "taxable_income": net, "tax": badan, "tax_without_facility": badan_full},
        "comparison": {"final": pph_final_annual, "badan": badan, "cheaper": cheaper, "saving": saving},
        "ppn": {"pkp": pkp, "rate_pct": 12, "output": ppn_out, "input": ppn_in,
                "payable": ppn_payable, "input_base": material_base},
        "pph21": {"rate_pct": 5, "payroll_base": payroll, "ptkp": PTKP_TK0,
                  "taxable": taxable21, "estimate": pph21},
        "pph23": {"rate_pct": 2, "rent_base": rent, "withhold_on_rent": pph23_withhold,
                  "credit_from_customers": pph23_credit},
        "monthly_obligations": obligations,
        "notes": [
            "Angka pajak adalah ESTIMASI berdasarkan data yang tercatat, bukan perhitungan pajak final.",
            "PPh Final UMKM 0,5% dihitung dari peredaran bruto (omzet), bukan dari laba.",
            "Fasilitas Pasal 31E memberi tarif efektif 11% untuk peredaran bruto sampai Rp4,8 miliar.",
            f"PPh 21 Rp{pph21:,.0f} — gaji {'di bawah' if pph21 == 0 else 'di atas'} PTKP Rp{PTKP_TK0:,.0f}.",
            "Konsultasikan dengan konsultan pajak / KPP untuk kepatuhan final.",
        ],
    }

File: frontend/src/pages/TaxCenter.js
import { useEffect, useState } from "react";
import { api, formatIDR } from "../lib/api";
import { StructuredAnswer } from "../components/StructuredAnswer";
import { InfoTip } from "../components/Common";
import { Card } from "../components/ui/card";
import { Button } from "../components/ui/button";
import { Switch } from "../components/ui/switch";
import { Skeleton } from "../components/ui/skeleton";
import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from "../components/ui/table";
import { Receipt, Sparkle, CheckCircle, Scales, Buildings, Percent } from "@phosphor-icons/react";

export default function TaxCenter() {
  const [pkp, setPkp] = useState(false);
  const [d, setD] = useState(null);
  const [ai, setAi] = useState(null);
  const [aiLoading, setAiLoading] = useState(false);

  useEffect(() => {
    setD(null);
    api.get(`/tax/summary?pkp=${pkp}`).then(({ data }) => setD(data)).catch(() => {});
  }, [pkp]);

  const advice = async () => {
    setAiLoading(true);
    try {
      const { data } = await api.get(`/ai/tax-advice?pkp=${pkp}`);
      setAi(data.ai);
    } catch (e) {} finally { setAiLoading(false); }
  };

  if (!d) return <Skeleton className="h-96 rounded-xl" />;
  const cmp = d.comparison;

  return (
    <div className="space-y-6" data-testid="tax-page">
      <div className="flex items-center justify-between flex-wrap gap-3">
        <div>
          <h1 className="font-display text-3xl font-bold tracking-tight flex items-center gap-2">
            <Receipt size={28} weight="fill" className="text-era-primary" /> Tax Center
          </h1>
          <p className="text-slate-500 text-sm mt-1">Estimasi kewajiban pajak: PPh Final UMKM, PPh Badan, PPN, PPh 21 & 23. Semua angka bersifat estimasi.</p>
        </div>
        <div className="flex items-center gap-3">
          <Card className="flex items-center gap-3 px-4 py-2 bg-white border border-era-border rounded-xl">
            <span className="text-sm text-slate-600">Status PKP</span>
            <Switch checked={pkp} onCheckedChange={setPkp} data-testid="pkp-toggle" />
            <span className={`text-xs font-semibold ${pkp ? "text-era-primary" : "text-slate-400"}`}>{pkp ? "PKP" : "Non-PKP"}</span>
          </Card>
          <Button onClick={advice} disabled={aiLoading} data-testid="tax-ai-advice" className="bg-era-primary hover:bg-emerald-800 rounded-md">
            <Sparkle size={16} weight="fill" className="mr-1.5" /> {aiLoading ? "Menganalisis..." : "Saran AI"}
          </Button>
        </div>
      </div>

      {/* Regime comparison */}
      <div className="grid grid-cols-1 lg:grid-cols-3 gap-6">
        <Card data-testid="pph-final-card" className={`p-6 rounded-xl border-2 ${cmp.cheaper === "final" ? "border-era-primary bg-emerald-50" : "border-era-border bg-white"}`}>
          <div className="flex items-center justify-between">
            <div className="flex items-center gap-2">
              <Scales size={20} weight="fill" className="text-era-primary" />
              <h3 className="font-display font-bold">PPh Final UMKM</h3>
            </div>
            {cmp.cheaper === "final" && <span className="text-[10px] font-bold uppercase bg-era-primary text-white rounded-full px-2 py-0.5">Termurah</span>}
          </div>
          <p className="text-xs text-slate-500 mt-1">0,5% × Peredaran Bruto (PP 55/2022)</p>
          <p className="font-display text-3xl font-extrabold text-era-primary tabular mt-3">{formatIDR(d.pph_final.annual)}</p>
          <p className="text-xs text-slate-500 mt-1">≈ {formatIDR(d.pph_final.monthly_avg)}/bulan</p>
        </Card>

        <Card data-testid="pph-badan-card" className={`p-6 rounded-xl border-2 ${cmp.cheaper === "badan" ? "border-era-primary bg-emerald-50" : "border-era-border bg-white"}`}>
          <div className="flex items-center justify-between">
            <div className="flex items-center gap-2">
              <Buildings size={20} weight="fill" className="text-era-secondary" />
              <h3 className="font-display font-bold">PPh Badan (Tarif Umum)</h3>
            </div>
            {cmp.cheaper === "badan" && <span className="text-[10px] font-bold uppercase bg-era-primary text-white rounded-full px-2 py-0.5">Termurah</span>}
          </div>
          <p className="text-xs text-slate-500 mt-1 flex items-center gap-1">
            Tarif efektif {d.pph_badan.effective_rate_pct}% {d.pph_badan.facility_31e && <span className="bg-blue-50 text-era-secondary rounded px-1.5 py-0.5">Fasilitas Ps.31E</span>}
          </p>
          <p className="font-display text-3xl font-extrabold text-era-secondary tabular mt-3">{formatIDR(d.pph_badan.tax)}</p>
          <p className="text-xs text-slate-400 mt-1">Tanpa fasilitas (22%): {formatIDR(d.pph_badan.tax_without_facility)}</p>
        </Card>

        <Card className="p-6 rounded-xl bg-white border border-era-border flex flex-col justify-center">
          <div className="flex items-center gap-1.5">
            <p className="text-sm text-slate-500">Potensi Penghematan</p>
            <InfoTip what="Selisih antara kedua skema pajak." why="Membantu memilih skema yang paling efisien secara legal." />
          </div>
          <p className="font-display text-3xl font-extrabold text-era-primary tabular mt-2">{formatIDR(cmp.saving)}</p>
          <p className="text-sm text-slate-600 mt-2 flex items-center gap-1">
            <CheckCircle size={16} weight="fill" className="text-era-success" />
            Rekomendasi: <span className="font-semibold">{cmp.cheaper === "final" ? "PPh Final 0,5%" : "PPh Badan"}</span>
          </p>
        </Card>
      </div>

      {ai && <StructuredAnswer a={ai} />}

      {/* Other taxes */}
      <div className="grid grid-cols-1 md:grid-cols-3 gap-6">
        <Card data-testid="ppn-card" className="p-6 bg-white border border-era-border rounded-xl">
          <h3 className="font-display font-bold flex items-center gap-1.5"><Percent size={18} weight="bold" className="text-era-secondary" /> PPN {pkp ? "(12%)" : ""}</h3>
          {pkp ? (
            <div className="mt-3 space-y-2 text-sm">
              <div className="flex justify-between"><span className="text-slate-500">PPN Keluaran</span><span className="tabular">{formatIDR(d.ppn.output)}</span></div>
              <div className="flex justify-between"><span className="text-slate-500">PPN Masukan</span><span className="tabular">{formatIDR(d.ppn.input)}</span></div>
              <div className="flex justify-between font-bold border-t border-slate-100 pt-2"><span>PPN Terutang</span><span className="tabular text-era-primary">{formatIDR(d.ppn.payable)}</span></div>
            </div>
          ) : (
            <p className="text-sm text-slate-500 mt-3">Status Non-PKP: tidak memungut PPN (omzet di bawah Rp4,8 M). Aktifkan toggle PKP untuk simulasi.</p>
          )}
        </Card>

        <Card data-testid="pph21-card" className="p-6 bg-white border border-era-border rounded-xl">
          <h3 className="font-display font-bold">PPh 21 (Gaji Karyawan)</h3>
          <div className="mt-3 space-y-2 text-sm">
            <div className="flex justify-between"><span className="text-slate-500">Dasar (Beban Gaji)</span><span className="tabular">{formatIDR(d.pph21.payroll_base)}</span></div>
            <div className="flex justify-between"><span className="text-slate-500">PTKP (TK/0)</span><span className="tabular">{formatIDR(d.pph21.ptkp)}</span></div>
            <div className="flex justify-between font-bold border-t border-slate-100 pt-2"><span>Estimasi PPh 21</span><span className="tabular text-era-primary">{formatIDR(d.pph21.estimate)}</span></div>
          </div>
          {d.pph21.estimate === 0 && <p className="text-xs text-slate-400 mt-2">Gaji di bawah PTKP → tidak ada PPh 21 terutang.</p>}
        </Card>

        <Card data-testid="pph23-card" className="p-6 bg-white border border-era-border rounded-xl">
          <h3 className="font-display font-bold">PPh 23 (Jasa & Sewa)</h3>
          <div className="mt-3 space-y-2 text-sm">
            <div className="flex justify-between"><span className="text-slate-500">Potong atas Sewa (2%)</span><span className="tabular">{formatIDR(d.pph23.withhold_on_rent)}</span></div>
            <div className="flex justify-between"><span className="text-slate-500">Dipotong pelanggan (kredit)</span><span className="tabular">{formatIDR(d.pph23.credit_from_customers)}</span></div>
          </div>
          <p className="text-xs text-slate-400 mt-2">Pelanggan badan dapat memotong 2% dari jasa Anda sebagai kredit pajak.</p>
        </Card>
      </div>

      {/* Monthly obligations */}
      <Card className="bg-white border border-era-border rounded-xl overflow-hidden">
        <div className="p-4 border-b border-slate-100 font-display font-bold">Kalender Kewajiban Pajak</div>
        <Table>
          <TableHeader><TableRow>
            <TableHead>Jenis Pajak</TableHead><TableHead>Frekuensi</TableHead>
            <TableHead>Jatuh Tempo</TableHead><TableHead className="text-right">Estimasi</TableHead>
          </TableRow></TableHeader>
          <TableBody>
            {d.monthly_obligations.map((o, i) => (
              <TableRow key={i} data-testid={`obligation-${o.type}`}>
                <TableCell className="font-medium">{o.tax}</TableCell>
                <TableCell><span className="text-xs bg-slate-100 rounded px-2 py-0.5">{o.freq}</span></TableCell>
                <TableCell className="text-slate-500 text-sm">{o.due}</TableCell>
                <TableCell className="text-right tabular font-semibold">{formatIDR(o.amount)}</TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </Card>

      <div className="bg-amber-50 border border-amber-100 rounded-xl p-4">
        <ul className="text-xs text-amber-800 space-y-1">
          {d.notes.map((n, i) => <li key={i}>• {n}</li>)}
        </ul>
      </div>
    </div>
  );
}
